### PR TITLE
feat(planner): support limit for new planner

### LINKED ISSUE #5240: Feature: `limit` for new planner
**Summary**

Description for this feature.

### PR DESCRIPTION
**Pull request by fkuner**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support limit for new planner.

The pr don't support expr in limit and offset now

## Changelog

- New Feature

## Related Issues

Fixes #5240

### CODE REVIEW

**Review by leiysky — APPROVED** [review]

File: query/src/sql/planner/binder/limit.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        offset: &Option<Expr<'a>>,
+        bind_context: &mut BindContext,
+    ) -> Result<()> {
+        let type_checker = TypeChecker::new(bind_context, self.ctx.clone());
+
+        let limit_cnt = match limit {
+            Some(Expr::Literal(x)) => Some(
+                type_checker
+                    .parse_literal(x, None)
+                    .unwrap()
```
Comment (BohuTANG) [inline query/src/sql/planner/binder/limit.rs:38]: Avoid using `unwrap`, let's return an error, or sets a default value.

File: query/src/sql/planner/binder/limit.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            Some(_) => {
+                return Err(ErrorCode::IllegalDataType("Unsupported limit type"));
+            }
+            None => None,
+        };
+
+        let offset_cnt = if let Some(Expr::Literal(x)) = offset {
+            type_checker
+                .parse_literal(x, None)
+                .unwrap()
```
Comment (BohuTANG) [inline query/src/sql/planner/binder/limit.rs:51]: ditto

File: query/src/sql/planner/binder/limit.rs (hunk 1 above)
Comment (fkuner) [reply query/src/sql/planner/binder/limit.rs:38]: ok

File: tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql [hunk 3]
```diff
@@ -91,4 +91,23 @@ select to_int(8);
...
+select number from numbers(100) order by number-2 asc limit 10, 10;
+select '=== Test limit with offset ===';
+select number from numbers(100) order by number asc limit 10 offset 10;
+select '==============================';
+select number from numbers(100) order by number/2 asc limit 10 offset 10;
+select '=== Test offset ===';
+select number from numbers(10) order by number asc offset 5;
+select '===================';
+select number from numbers(10) order by number+number asc offset 5;
+select number from numbers(10000000000000) limit 1;
```
Comment (sundy-li) [inline tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql:111]: We did not push down the limit to data source in v2 planner. So let's remove this heavy test.

Comment (fkuner) [reply tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql:111]: got it

**Comment by sundy-li** [comment]
> Seems to support expr in `limit`(such as limit 1 + 1, or limit 5 / 2) need to change the logic of `LimitTransform`, any thought? @leiysky @sundy-li

It might be done in expression constant folding rule ?

**Comment by xudong963** [comment]
> It might be done in expression constant folding rule ?

Oh, we can support it after we add constant folading rule in the new optimizer.

File: query/src/sql/planner/binder/limit.rs [hunk 4]
```diff
@@ -32,25 +32,43 @@ impl<'a> Binder {
...
 
         let limit_cnt = match limit {
-            Some(Expr::Literal(x)) => Some(
-                type_checker
-                    .parse_literal(x, None)
-                    .unwrap()
-                    .as_u64()
-                    .unwrap() as usize,
-            ),
+            Some(Expr::Literal(x)) => {
```
Comment (BohuTANG) [inline query/src/sql/planner/binder/limit.rs:35]:
```suggestion
let limit_cnt = match limit {
            Some(Expr::Literal(x)) => {
                let data = type_checker.parse_literal(x, None)?.as_u64()?;
                Some(data as usize)
            }
            Some(_) => {
                return Err(ErrorCode::IllegalDataType("Unsupported limit type"));
            }
            None => None,
        };

        let offset_cnt = if let Some(Expr::Literal(x)) = offset {
            let data = type_checker.parse_literal(x, None)?.as_u64()?;
            data as usize
        } else {
            0
        };
```

Comment (fkuner) [reply query/src/sql/planner/binder/limit.rs:35]: thanks.

**Review by BohuTANG — APPROVED** [review]

**Pull request merged**